### PR TITLE
New flag "-ssl-upstream-insecure-skip-validation"

### LINKED ISSUE #75: Skip validation of upstream certificate
Even with `-ssl-insecure-skip-verify` it is not possible to use a self-signed certificate for the upstream.

## Expected Behavior

Upstream with self-signed certificate (such as default kubernetes dashboard deployment) should work if oauth2-proxy is configured with `-ssl-insecure-skip-verify`.

## Current Behavior

Oauth2-proxy still tries to verify the certificate and fails with the following error message:

```
reverseproxy.go:395: http: proxy error: x509: certificate has expired or is not yet valid
```

## Possible Solution

This issue was reported [here](https://github.com/bitly/oauth2_proxy/issues/403) in the bitly/oauth2_proxy repository and has a pull request with a proposed solution [here](https://github.com/bitly/oauth2_proxy/pull/651).

## Steps to Reproduce (for bugs)

I have followed [this guide](https://blog.heptio.com/on-securing-the-kubernetes-dashboard-16b09b1b7aca) but modified it a bit.

1. Install [minikube](https://kubernetes.io/docs/setup/minikube/), kubectl and [helm](https://github.com/helm/helm#install)
2. Clone [this repo](https://github.com/lentzi90/oauth2-proxy-experiment)
3. Checkout the branch `self-signed-dashboard`
4. Follow the quick start
5. Check the logs of the oauth2-proxy pod

## Context

The default dashboard deployment uses self-signed certificates and this is better than no TLS at all.
I think this is a common use case, especially in a kubernetes environment.

## Your Environment

Versions used:

- Kubernetes v1.13.3
- Minikube v0.34.1
- Oauth2-proxy chart 0.9.1
- Oauth2-proxy 3.1.0

### PR DESCRIPTION
**Pull request by jansinger**
## Motivation and Context

Upstream with self-signed certificate (such as default kubernetes dashboard deployment) should work if oauth2-proxy.

Fixes #75 

## How Has This Been Tested?

Setup an oauth2_proxy for a Kubernetes dashboard as described in Issue #75, produced the errors as mentioned without the new flag, works setting the flag.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by steakunderscore** [comment]
Hi @jansinger :wave: Thanks for your PR. Did you look in to why the tests are failing?

**Comment by jansinger** [comment]
Hi @steakunderscore, sorry, I just missed to push the updated test after changing the method signature of NewReverseProxy(). It's fixed now.

**Comment by jansinger** [comment]
Can anyone give me a hint why this is failing now when trying to download the mods?

**Comment by syscll** [comment]
It could be due to Travis just timing out as this is something I've experienced before with go modules. I've restarted the build, but if this issue persists then I'll investigate further.

**Comment by syscll** [comment]
Just a thought on this, why do we need to add a new flag? Can we not just pass the `-ssl-insecure-skip-verify` flag to the `NewReverseProxy` func instead? As I assume needing to skip verification will always require both `-ssl-insecure-skip-verify` and `-ssl-upstream-insecure-skip-verify` being set to true.

**Comment by jansinger** [comment]
We want to validate the SSL certificate of our OIDC provider, which has an official certificate and can be validated and treated as secure. But the upstreams in our Kubernetes Cluster have self signed certs, so validation need to be skipped for them.
If all agree that it makes sense to always skip validation of SSL certificates for both providers and upstreams, I can change it accordingly. We will then continue to use our own fork to have this seperated.

**Comment by JoelSpeed** [comment]
Personally I would keep them separate. There are two different "sets" of certificates and I can see valid reason for wanting to validate one set and not validate the others

**Comment by syscll** [comment]
Yeah after reading this again it definitely makes sense to keep them separate!

**Comment by syscll** [comment]
Are you happy with this @JoelSpeed @steakunderscore?

File: CHANGELOG.md [hunk 1]
```diff
@@ -36,10 +36,13 @@
 
 ## Changes since v3.2.0
 
+- [#234](https://github.com/pusher/oauth2_proxy/pull/234) Added option `-ssl-upstream-insecure-skip-validation` to skip validation of upstream SSL certificates (@jansinger)
 - [#224](https://github.com/pusher/oauth2_proxy/pull/224) Check Google group membership using hasMember to support nested groups and external users (@jpalpant)
 - [#231](https://github.com/pusher/oauth2_proxy/pull/231) Add optional group membership and email domain checks to the GitLab provider (@Overv)
 - [#178](https://github.com/pusher/outh2_proxy/pull/178) Add Silence Ping Logging and Exclude Logging Paths flags (@kskewes)
 - [#209](https://github.com/pusher/outh2_proxy/pull/209) Improve docker build caching of layers (@dekimsey)
+- [#178](https://github.com/pusher/oauth2_proxy/pull/178) Add Silence Ping Logging and Exclude Logging Paths flags (@kskewes)
```
Comment (JoelSpeed) [inline CHANGELOG.md]: There's something up with a merge here, looks like these are duplicated?

**Comment by syscll** [comment]
Thank you @jansinger and congrats on your first contribution! :tada:

**Pull request merged**